support/cas-server-support-x509-core/src/main/java/org/apereo/cas/adaptors/x509/authentication/principal/X509UPNExtractor.java

```diff
@@ -0,0 +1,124 @@
+        if (id != null && UPN_OBJECTID.equals(id.getId())) {
+            val obj = (ASN1TaggedObject) seq.getObjectAt(1);
+            val primitiveObj = obj.getObject();
+
+            val func = FunctionUtils.doIf(Predicates.instanceOf(ASN1TaggedObject.class),
+                () -> ASN1TaggedObject.getInstance(primitiveObj).getObject(),
+                () -> primitiveObj);
+            val prim = func.apply(primitiveObj);
+
+            if (prim instanceof ASN1OctetString) {
+                return new String(((ASN1OctetString) prim).getOctets(), StandardCharsets.UTF_8);
+            }
+            if (prim instanceof ASN1String) {
+                return ((ASN1String) prim).getString();
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Get alt name seq.
+     *
+     * @param sanItem subject alternative name value encoded as a two elements List with elem(0) representing object id and elem(1)
+     *                representing object (subject alternative name) itself.
+     * @return ASN1Sequence abstraction representing subject alternative name or null if the passed in
+     * List doesn't contain at least two elements.
+     * as expected to be returned by implementation of {@code X509Certificate.html#getSubjectAlternativeNames}
+     * @see <a href="http://docs.oracle.com/javase/7/docs/api/java/security/cert/X509Certificate.html#getSubjectAlternativeNames()">
+     * X509Certificate#getSubjectAlternativeNames</a>
+     */
+    private static ASN1Sequence getAltnameSequence(final List sanItem) {
+        //Should not be the case, but still, a extra "safety" check
+        if (sanItem.size() < 2) {
+            LOGGER.error("Subject Alternative Name List does not contain at least two required elements. Returning null principal id...");
+        }
+        val itemType = (Integer) sanItem.get(0);
+        if (itemType == 0) {
+            val altName = (byte[]) sanItem.get(1);
+            return getAltnameSequence(altName);
+        }
+        return null;
+    }
+
+    /**
+     * Get alt name seq.
+     *
+     * @param sanValue subject alternative name value encoded as byte[]
+     * @return ASN1Sequence abstraction representing subject alternative name
+     * @see <a href="http://docs.oracle.com/javase/7/docs/api/java/security/cert/X509Certificate.html#getSubjectAlternativeNames()">
+     * X509Certificate#getSubjectAlternativeNames</a>
+     */
+    private static ASN1Sequence getAltnameSequence(final byte[] sanValue) {
+        try (val bInput = new ByteArrayInputStream(sanValue);
+             val input = new ASN1InputStream(bInput)) {
+            return ASN1Sequence.getInstance(input.readObject());
+        } catch (final IOException e) {
+            LOGGER.error("An error has occurred while reading the subject alternative name value", e);
+        }
+        return null;
+    }
+
+    public static String extractUPNString(final X509Certificate certificate) throws CertificateParsingException {
```

Comment:
Add javadoc please, and perhaps a few additional logging statements.

Reply:
Let me know if that is too much logging. The trace logging could be useful if someone weren't getting a UPN that they thought they should because it logs every SAN that Java returns.